### PR TITLE
Release v2.0

### PR DESCRIPTION
**Pull request by abertschi**
- introduce new endpoints https://pccweb.api.post.ch/secure/api/mobile/v1 #22 #21

### CODE REVIEW

**Comment by wp99cp** [comment]
Have you testet sending postcards with an umlaut in the address field? Today I revised (I'm surprised that the Swiss Post can handle this) a card with wrong encoding of umlauts in the address field. However, I am not 100% sure that the error is not in my code.

**Comment by abertschi** [comment]
Can you elaborate what you did and what you mean by wrong encoding of umlauts? I encode characters in the address fields as ascii with xmlcharrefreplace. These fields were embedded in an svg in the legacy endpoints. The same encoding is also used for the new endpoints. It is possible that the same encoding can cause troubles with the new endpoints.

**Comment by wp99cp** [comment]
> Can you elaborate what you did and what you mean by wrong encoding of umlauts? I encode characters in the address fields as ascii with xmlcharrefreplace. These fields were embedded in an svg in the legacy endpoints. The same encoding is also used for the new endpoints. It is possible that the same encoding can cause troubles with the new endpoints. I haven't received a postcard with umlauts yet using the new endpoints.

My address contains some umlauts, for example the "ü" in Zurich. On the card, these umlauts were coded as HTML characters ("&amp;"). Unfortunately, I did not use any umlauts in the text, so I cannot say whether there are also umlauts there. But I'm not sure if it's your code or if the error is mine. I am currently developing an automatic mailing system that sends out a card every week, the data is stored in a pandas dataframe or as a CSV and is saved in the script. However, it certainly won't hurt if we test this again. A new postcard from me is already on its way to my place, but this takes some time, as we all know... Testing these things out is slowly.

**Comment by abertschi** [comment]
> My address contains some umlauts, for example the "ü" in Zurich. On the card, these umlauts were coded as HTML characters (&amp;). Unfortunately, I did not use any umlauts in the text, so I cannot say whether there are also umlauts there. But I'm not sure if it's your code or if the error is mine. I am currently developing an automatic mailing system that sends out a card every week, the data is stored in a pandas dataframe or as a CSV and is saved in the script. However, it certainly won't hurt if we test this again. A new postcard from me is already on its way to my place, but this takes some time, as we all know... Testing these things out is slowly.

The character conversion with & is employed in the wrapper. This was necessary because in the legacy wrapper the address data was directly embedded in svg which did not support umlauts. Good to know! :) I'll fix this then.

> Unfortunately, I did not use any umlauts in the text,

The body text is rendered and uploaded as an image so this won't cause problems.

**Comment by abertschi** [comment]
- [ ] test address fields with non-ascii characters, emojis

**Comment by wp99cp** [comment]
> * [ ]  test address fields with non-ascii characters, emojis

With your update, the umlauts now work without a problem, thank you.

File: postcard_creator/postcard_img_util.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            else:
+                # does fit
+                size_l = size + 1
+
+        return last_size, last_line_w
+
+    def center_y(lines, font_h):
+        tot_height = len(lines) * font_h
+        if tot_height < text_canvas_h:
+            return (text_canvas_h - tot_height) / 2
```
Comment (wp99cp) [inline postcard_creator/postcard_img_util.py:144]: Depending on the image resolution I get the following error message:

```
 File "/home/pi/postCards/postcard_creator/postcard_creator.py", line 96, in wrapped
    return func(*args, **kwargs)
  File "/home/pi/postCards/postcard_creator/postcard_creator_swissid.py", line 117, in send_free_card
    img_text_base64 = base64.b64encode(self.create_text_cover(postcard.message)).decode('ascii')
  File "/home/pi/postCards/postcard_creator/postcard_creator_swissid.py", line 153, in create_text_cover
    return create_text_image(msg, image_export=True)
  File "/home/pi/postCards/postcard_creator/postcard_img_util.py", line 163, in create_text_image
    embedded_color=True)
  File "/home/pi/.local/lib/python3.7/site-packages/PIL/ImageDraw.py", line 379, in text
    draw_text(ink)
  File "/home/pi/.local/lib/python3.7/site-packages/PIL/ImageDraw.py", line 361, in draw_text
    self.im.paste(color, coord + coord2, mask)
TypeError: integer argument expected, got float
```
I suggest to use an interger division here:

```
return (text_canvas_h - tot_height) // 2
```

File: postcard_creator/postcard_img_util.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    font = load_font(size)
+    font_w, font_h = font.getsize(text)
+    lines = textwrap.wrap(text, width=line_w)
+    text_y_start = center_y(lines, font_h)
+
+    canvas = Image.new('RGB', (text_canvas_w, text_canvas_h), text_canvas_bg)
+    draw = ImageDraw.Draw(canvas)
+    for line in lines:
+        width, height = font.getsize(line)
+        draw.text(((text_canvas_w - width) / 2, text_y_start), line,
```
Comment (wp99cp) [inline postcard_creator/postcard_img_util.py:160]: Depending on the image resolution I get the following error message:

```
 File "/home/pi/postCards/postcard_creator/postcard_creator.py", line 96, in wrapped
    return func(*args, **kwargs)
  File "/home/pi/postCards/postcard_creator/postcard_creator_swissid.py", line 117, in send_free_card
    img_text_base64 = base64.b64encode(self.create_text_cover(postcard.message)).decode('ascii')
  File "/home/pi/postCards/postcard_creator/postcard_creator_swissid.py", line 153, in create_text_cover
    return create_text_image(msg, image_export=True)
  File "/home/pi/postCards/postcard_creator/postcard_img_util.py", line 163, in create_text_image
    embedded_color=True)
  File "/home/pi/.local/lib/python3.7/site-packages/PIL/ImageDraw.py", line 379, in text
    draw_text(ink)
  File "/home/pi/.local/lib/python3.7/site-packages/PIL/ImageDraw.py", line 361, in draw_text
    self.im.paste(color, coord + coord2, mask)
TypeError: integer argument expected, got float
```

I suggest to use an interger division here:

```
draw.text(((text_canvas_w - width) // 2, text_y_start), line,
```

Comment (abertschi) [reply postcard_creator/postcard_img_util.py:160]: nice catch! Yes, LGTM

**Comment by wp99cp** [comment]
I think after merging #25, v2.0 seems to be stable enough to be released. I've sent a hand full postcards, without any other issues, than the ones we have addressed. What do you think?

**Pull request merged**